hail/src/main/scala/is/hail/expr/ir/functions/StringFunctions.scala

```diff
@@ -96,11 +96,13 @@ object StringFunctions extends RegistryFunctions {
   def registerAll(): Unit = {
     val thisClass = getClass
 
-    registerCode("length", TString, TInt32, null) { case (r: EmitRegion, rt, (sT: PString, s: Code[Long])) =>
+    registerCode("length", TString, TInt32, (pt: PType) => pt) { case (r: EmitRegion, rt, (sT: PString, s: Code[Long])) =>
       asm4s.coerce[String](wrapArg(r, sT)(s)).invoke[Int]("length")
     }
 
-    registerCode("substring", TString, TInt32, TInt32, TString, null) {
+    registerCode("substring", TString, TInt32, TInt32, TString, {
+      (sT: PType, startT: PType, endT: PType) => PCanonicalString(sT.required && (startT.required || endT.required))
```

Comment:
sure

Reply:
I misunderstood, I thought that you had wanted me to implement requiredeness in registerCode, but not to worry about getting it exactly right because `registerCode` would override (I had read into this that we would at some point give the registerCode callers control of their requiredeness, to allow finer-grained decisions). Will change.